File: LeetCode/047 Permutations II.py
# Backtracking

# Given a collection of numbers that might contain duplicates, return all possible unique permutations.

# Example:

# Input: [1,1,2]
# Output:
# [
#   [1,1,2],
#   [1,2,1],
#   [2,1,1]
# ]

class Solution:
    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        def allPerms(elements):
            if len(elements) < 2:
                yield elements
            else:
                for perm in allPerms(elements[1:]):
                    for i in range(len(perm) + 1):
                        yield perm[:i] + elements[:1] + perm[i:]
        output = []
        for x in allPerms(nums):
            if x not in output:
                output.append(x)
        return output
        

File: LeetCode/047 Permutations II.java
// Backtracking

// Given a collection of numbers that might contain duplicates, return all possible unique permutations.

// Example:

// Input: [1,1,2]
// Output:
// [
//   [1,1,2],
//   [1,2,1],
//   [2,1,1]
// ]

class Solution {
    public List<List<Integer>> permuteUnique(int[] nums) {
        Set<List<Integer>> solutions = new HashSet<>();
        permute(nums, 0, new boolean[nums.length], solutions, new ArrayList<>());
        return new ArrayList<>(solutions);
    }
    
    private void permute(int[] array, int index, boolean[] used, Set<List<Integer>> solutions, List<Integer> list) {
        if (index == array.length) {
            solutions.add(new ArrayList<>(list));
            return;
        }
        for (int i = 0; i < array.length; i++) {
            if (used[i] == false) {
                list.add(array[i]);
                used[i] = true;
                permute(array, index + 1, used, solutions, list);
                used[i] = false;
                list.remove(list.size() - 1);
            }
        }
    }
}